sips/committee_consensus.md

```diff
@@ -0,0 +1,486 @@
+
+The current structure that we have in code can be kept with small changes.
+
+```go
+type PartialSignatureMessages struct {
+	Type     PartialSigMsgType
+	Slot     phase0.Slot
+	Messages []*PartialSignatureMessage
+}
+
+// PartialSignatureMessage is a msg for partial Beacon chain related signatures (like partial attestation, block, randao sigs)
+type PartialSignatureMessage struct {
+	PartialSignature Signature `ssz-size:"96"` // The Beacon chain partial Signature for a duty
+	SigningRoot      [32]byte  `ssz-size:"32"` // the root signed in PartialSignature
+	Signer         OperatorID
+	ValidatorIndex phase0.ValidatorIndex	// addition
+}
+```
+
+We must have the following flow:
+1. Calculate `expectedRootsAndBeaconObjects()`. Use the `duty` objects to reconstruct the proper beacon data objects (i.e. `Attestation`).
+2. The above calculation can be used to create a mapping of `ValidatorIndex` to `root`, and `root` to `beaconObject`.
+3. When processing the messages find all the roots that have quorums for a certain validator, mark them, and submit corresponding beacon data to beacon chain.
+4. For the next message received attempt to complete quorum for other roots.
+
+
+#### Happy Flow
+
+1. `Committee` receives duties that match a certain slot. Then `StartDuty` and filter the validators who have a higher `HighestAttestingSlot`. Initialize a `CommitteeRunner` for the relevant Validators and slot. For each validator that we start, mark its old beacon duties as stopped.
+```go
+// StartDuty starts a new duty for the given slot
+func (c *Committee) StartDuty(duty *types.CommitteeDuty) error {
+	if _, exists := c.Runners[duty.Slot]; exists {
+		return errors.New(fmt.Sprintf("CommitteeRunner for slot %d already exists", duty.Slot))
+	}
+	c.Runners[duty.Slot] = c.CreateRunnerFn()
+	validatorToStopMap := make(map[spec.Slot]types.ValidatorPK)
+	// Filter old duties based on highest attesting slot
+	duty, validatorToStopMap, c.HighestAttestingSlotMap = FilterCommitteeDuty(duty, c.HighestAttestingSlotMap)
+	// Stop validators with old duties
+	c.stopDuties(validatorToStopMap)
+	c.updateAttestingSlotMap(duty)
+	return c.Runners[duty.Slot].StartNewDuty(duty)
+}
+
+// FilterCommitteeDuty filters the committee duty. It returns the new duty, the validators to stop and the highest attesting slot map
+func FilterCommitteeDuty(duty *types.CommitteeDuty, slotMap map[types.ValidatorPK]spec.Slot) (
+	*types.CommitteeDuty,
+	map[spec.Slot]types.ValidatorPK,
+	map[types.ValidatorPK]spec.Slot) {
+	validatorsToStop := make(map[spec.Slot]types.ValidatorPK)
+
+	for i, beaconDuty := range duty.BeaconDuties {
+		validatorPK := types.ValidatorPK(beaconDuty.PubKey)
+		slot, exists := slotMap[validatorPK]
+		if exists {
+			if slot < beaconDuty.Slot {
+				validatorsToStop[beaconDuty.Slot] = validatorPK
+				slot = beaconDuty.Slot
+			} else { // else don't run duty with old slot
+				duty.BeaconDuties[i] = nil
+			}
+		}
+	}
+	return duty, validatorsToStop, slotMap
+}
```

Comment:
won't this stop attester duties after one slot for validators who currently have a sync duty?

Reply:
Yes, I've already fixed this issue in implementation.
@GalRogozinski